ls1intum/Artemis
# TextExercise creation/update for exam-mode

### Checklist
- [x] Client: I added `authorities` to all new routes

### Motivation and Context
TextExercise creation/update must be adapted for exam-mode. We need to ensure that Course and ExerciseGroup are never set simultaneously.

### Description
Client:
- [x] Create new client routes for text-exercise update and creation
For creation: ':examId/exerciseGroups/:groupId/text-exercises/new'
For update: ':examId/exerciseGroups/:groupId/text-exercises/:exerciseId/edit'
- [x] Load text exercise module
- [x] Extend TextExerciseResolver by fetching of the exerciseGroup for the new routes
- [x] Deactivate categories in exam mode, as this relies on a Course
- [x] Deactivate notificationInput for examMode

Server:
- [x] Created ExerciseGroupService
- [x] Deactivate Tutor/Student notification on exercise creation/update
- [x] Add additional authorization check to GET text-exercise route. Only instructors and admins should be allowed to access an exam text exercise
- [x] Deactivate access to exercise details for exam exercises. Need another case distinction here or another route.  

### Steps for Testing
_Ensure that the existing text exercise creation/update functionality still works_

1. Create a regular text exercise
--> Exercise was created
2. Update the exercise
--> Exercise was updated

_Ensure that the new routes for text exercise work_
**This can only be tested locally as long as exam and exerciseGroup creation is not merged.**
1. Insert a new exam linked to a course into the database
2. Insert a new exerciseGroup linked to the exam into the database
--> Use course, exam and exerciseGroup id for the following routes
3. Log in as instructor or admin
4. Navigate to course-management/:courseId/exams/:examId/exerciseGroup/:groupId/text-exercises/new
--> Create a new text exercise and remember the created id
5. Navigate to course-management/:courseId/exams/:examId/exerciseGroup/:groupId/text-exercises/:exerciseId/edit
--> The text exercise should be displayed correctly
6. Edit the text exercise
7. Navigate again to course-management/:courseId/exams/:examId/exerciseGroup/:groupId/text-exercises/:exerciseId/edit
--> Ensure that the exercise was updated correctly

_Ensure correct authorization for exam exercises_
1. Log in as a tutor or student
2. Navigate to course-management/:courseId/exams/:examId/exerciseGroup/:groupId/text-exercises/new
--> You should not be able to access the route
4. Use the id of an exam exercise for the following steps
5. Navigate to course-management/:courseId/exams/:examId/exerciseGroup/:groupId/text-exercises/:exerciseId/edit
--> You should not be able to access the route
6. Navigate to /courses/:courseId/exercises/:exerciseId
--> You should not be able to access the route
7. Navigate to /course-management/:courseId/text-exercises/:exerciseId
--> You should not be able to access the route

_Ensure correct authorization for normal exercises as tutor_
1. Log in as a tutor
2. Use the id of a normal text exercise for the following steps
4. Navigate to /courses/:courseId/exercises/:exerciseId
--> The page should load properly
5. Navigate to /course-management/:courseId/text-exercises/:exerciseId
--> The page should load properly

_Ensure correct authorization for normal exercises as student_
1. Log in as a student
2. Use the id of a normal text exercise
4. Navigate to /courses/:courseId/exercises/:exerciseId
--> The page should load properly
